lib/src/clojure_lsp/dep_graph.clj

```diff
@@ -0,0 +1,344 @@
+;; namespace, falsy otherwise. May be absent if the namespace isn't required by
+;; other namespaces.
+
+;; :from-langs is a multiset of the langs of the :namespace-usages or files (if the
+;; :namespace-usages don't have langs) that use this namespace. May be either
+;; empty or absent if the namespace isn't required by other namespaces.
+
+;; file-meta-item =
+;; {(:namespaces #{ns*})?
+;;  :internal? boolean
+;;  :langs #{lang+}}
+
+;; :namespaces is a set of the namespaces defined by this file. May be either
+;; empty or absent if the file doesn't define any namespaces.
+
+;; :internal? is whether this file is internal to the project.
+
+;; :langs is a set of the langs used by this file.
+
+;; file-meta = {(filename file-meta-item)*}
+;; dep-graph = {(ns dep-graph-item)*}
+;; db = {':dep-graph' dep-graph
+;;       ':file-meta' file-meta}
+
+(comment
+  ;; Example
+  (require '[clojure-lsp.db :as db])
+  (-> @db/db*
+      :dep-graph
+      (select-keys '[clojure-lsp.main
+                     clojure.tools.cli]))
+  '{clojure-lsp.main  {:dependents     {clojure-lsp.main-test 1}
+                       :aliases        {main 1}
+                       :from-internal? true
+                       :from-langs     {:clj 1}
+                       :dependencies   {borkdude.dynaload        1
+                                        clojure-lsp.internal-api 1
+                                        clojure-lsp.kondo        1
+                                        clojure-lsp.server       1
+                                        clojure-lsp.shared       1
+                                        clojure.core             1
+                                        clojure.edn              1
+                                        clojure.java.io          1
+                                        clojure.string           1
+                                        clojure.tools.cli        1
+                                        pod.clojure-lsp.api      1}
+                       :files          #{"~/code/clojure-lsp/cli/src/clojure_lsp/main.clj"}
+                       :internal?      true}
+    clojure.tools.cli {:dependents     {clj-depend.main                                    1
+                                        cljfmt.main                                        1
+                                        clojure-lsp.main                                   1
+                                        clojure.tools.deps.alpha.script.generate-manifest2 1
+                                        clojure.tools.deps.alpha.script.make-classpath2    1
+                                        clojure.tools.deps.alpha.script.print-tree         1
+                                        clojure.tools.deps.alpha.script.resolve-tags       1
+                                        kaocha.runner                                      1}
+                       :aliases        {cli 7, nil 1}
```

Comment:
Should we have some way to differ `[require foo]` from `[require foo :refer :all]` ? would both be just `nil`?

Reply:
Yes, both would be `nil`. So far there hasn't been any need to distinguish situations that use `:refer`. I think that's mostly because this code is about namespace definitions and usages, not var definitions and usages.

Actually, the presence of the `nil`s is mostly an accident. Allowing `nil` to be present makes the dep graph maintenance code a little simpler. None of the other code _needs_ to know which namespaces are required without an alias. Avoiding the `nil`s would be possible, but would add an `if` to handle an edge case we don't actually care about.

Reply:
yeah, makes sense, looks good for now